pkg/vulnsrc/vulnsrc.go

```diff
@@ -137,39 +137,58 @@ type Optimizer interface {
 }
 
 type fullOptimizer struct {
-	dbc db.Operation
+	dbConfig db.VulnOperation
+	dbOp     db.Operation
 }
 
 func (o fullOptimizer) Optimize() error {
-	err := o.dbc.ForEachSeverity(func(tx *bolt.Tx, cveID string, _ types.Severity) error {
-		severity, title, description, references := vulnerability.GetDetail(cveID)
-		vuln := types.Vulnerability{
-			Title:       title,
-			Description: description,
-			Severity:    severity.String(),
-			References:  references,
-		}
-		if err := o.dbc.PutVulnerability(tx, cveID, vuln); err != nil {
-			return xerrors.Errorf("failed to put vulnerability: %w", err)
-		}
-		return nil
+	err := o.dbOp.ForEachSeverity(func(tx *bolt.Tx, cveID string, _ types.Severity) error {
+		return o.fullOptimize(cveID, tx)
 	})
 	if err != nil {
 		return xerrors.Errorf("failed to iterate severity: %w", err)
 	}
 
-	if err := o.dbc.DeleteSeverityBucket(); err != nil {
+	if err := o.dbOp.DeleteSeverityBucket(); err != nil {
 		return xerrors.Errorf("failed to delete severity bucket: %w", err)
 	}
 
-	if err := o.dbc.DeleteVulnerabilityDetailBucket(); err != nil {
+	if err := o.dbOp.DeleteVulnerabilityDetailBucket(); err != nil {
 		return xerrors.Errorf("failed to delete vulnerability detail bucket: %w", err)
 	}
 
 	return nil
 
 }
 
+// TODO: Until we can dependency inject, we need to monkey patch sadly
+var (
+	getDetailFunc = vulnerability.GetDetail
+)
+
+func (o fullOptimizer) fullOptimize(cveID string, tx *bolt.Tx) error {
+	severity, title, description, references := getDetailFunc(cveID)
+	vuln := types.Vulnerability{
+		Title:       title,
+		Description: description,
+		Severity:    severity.String(),
+		References:  references,
+	}
+
+	// assign vendor severities
+	vs := make(types.VendorSeverity)
+	vd, _ := o.dbConfig.GetVulnerabilityDetail(cveID)
+	for vendor, detail := range vd {
```

Comment:
As we discussed before, we should not use all vendor severities because some vendors may use the same severity as NVD.

Reply:
Also, we have to investigate the severity of public databases to detect vulnerabilities of libraries. We have the list here.
https://github.com/aquasecurity/trivy-db/blob/master/Makefile#L69-L73

Reply:
OK I updated it here 7270ae49ccc114d62b3109b3b5e5b34c441db0d5

Reply:
> As we discussed before, we should not use all vendor severities because some vendors may use the same severity as NVD.

Even if they do store the same as NVD, I'd prefer keeping the logic simple rather than removing them. To me removing them feels premature optimization. As we can see, even with keeping them the size is hardly any different.


> Also, we have to investigate the severity of public databases to detect vulnerabilities of libraries. We have the list here.
> https://github.com/aquasecurity/trivy-db/blob/master/Makefile#L69-L73

I'm not sure what you imply by this. Could you give an example of how to use these databases? Is it part of this PR or can we move it in a different PR?

Reply:
They also have severities. If we detect vulnerabilities of npm and show the severity of Red Hat, it is weird. We have to display the severity of the npm database.
https://github.com/nodejs/security-wg/blob/master/vuln/npm/208.json#L20

Reply:
>Even if they do store the same as NVD, I'd prefer keeping the logic simple rather than removing them.

I got it. We will keep as is.

Reply:
> They also have severities. If we detect vulnerabilities of npm and show the severity of Red Hat, it is weird. We have to display the severity of the npm database.
> https://github.com/nodejs/security-wg/blob/master/vuln/npm/208.json#L20

As discussed offline, you can see that with this commit it now favours the priority that the vendor score has: https://github.com/aquasecurity/trivy-db/pull/32/commits/99cdc188683d9f1b7e659ab2f9e1fbe669f683eb 

An example output of a CVE 
<img width="1539" alt="image" src="https://user-images.githubusercontent.com/1254783/77013758-fac6d180-692d-11ea-9589-f3fe59496fc0.png">